### ISSUE TITLE
用户自定义规则无效

### ISSUE THREAD

**Issue by derekzone** [shadowsocks/ShadowsocksX-NG#208]
用户自定义规则无效，是我添加的规则有问题吗？
||ip138.com

**Comment by qiuyuzhou** [comment]
继承原版的pac实现，其实我也不很清楚该怎么写规则。

**Comment by maddie** [comment]
@derekzone  直接写domain即可

如
```
google.com
youtube.com
cdn.instagram.com
```

**Comment by derekzone** [comment]
@maddie 试了下，虽然通知“PAC has been updated by user rules”但还是走原来的PAC,我拿的baidu.com测试，搜索IP还是显示的本地IP。

**Comment by maddie** [comment]
@derekzone 可以试试切换PAC/Global mode，或者Turn Shadowsocks on/off，或者重新打开ShadowsocksX-NG

1.3.2版本貌似会有PAC不生效的问题，1.4.0 alpha中的 cb4e6d7 有修改针对PAC文件的检测，可以更新试试

**Comment by weithink** [comment]
@maddie 如果想某个域名下的所有子域名都走代理的话，要怎么写

**Comment by qiuyuzhou** [comment]
https://adblockplus.org/en/filter-cheatsheet

**Comment by LemonJeer** [comment]
@qiuyuzhou Followed all the instructions on that, but the customized PAC is not working at all. Seems many people meet the same issue......

**Comment by reggiezhang** [comment]
执行以下"Update PAC from GFW list", 重新生成pac文件。然后可以访问http://127.0.0.1:1089/proxy.pac看看
Suppose should execute "Update PAC from GFW list" again to regenerate the PAC file, then you can verify it through http://127.0.0.1:1089/proxy.pac

**Comment by biggbuddy** [comment]
@maddie 让某一个网站不走代理应该怎么写?? 我现在是国内的如知乎, 163这样的网站都会走代理...............................

**Comment by maddie** [comment]
@biggbuddy 我也不太清楚，不过可以试试： https://adblockplus.org/en/filter-cheatsheet#exceptions1

**Comment by xhlwill** [comment]
@maddie thx 经测有效

**Comment by xhlwill** [comment]
FYI https://github.com/gfwlist/gfwlist

**Comment by biggbuddy** [comment]
@maddie 经测试使用以下方式可以, 但设置后需要重启shadowsocks-ng

```
@@alibaba.com
```

**Comment by ahlixinjie** [comment]
还是不行

**Comment by yangfeicheung** [comment]
可以试试我更改版本的规则，比较简单，直接*匹配即可,原版的规则有点拗口，看的头疼

*.somesite.com 就能匹配这个站点

**Comment by crazyyi** [comment]
@biggbuddy 说的没错，一定要重启服务。

**Comment by cuimingda** [comment]
折腾了好半天才找到这个帖子，我就说规则为什么一直不生效，关闭重新打开就生效了。
还有那个gfwlist，其实挺冗余的，有很多网站根本用不到，甚至都已经关闭了，如果可以自己配置更新源就好了。

**Comment by zs8861** [comment]
> @derekzone 直接写domain即可
> 
> 如
> 
> ```
> google.com
> youtube.com
> cdn.instagram.com
> ```

yes, you're right, I write the website in the rule list, and close the SS, then restart it, that's OK, nice :)

**Comment by ericliang** [comment]
> @maddie 试了下，虽然通知“PAC has been updated by user rules”但还是走原来的PAC,我拿的baidu.com测试，搜索IP还是显示的本地IP。

Don't use baidu.com for the test since it is in the whitelist for exclusion which will override your rule.

**Comment by Nomia** [comment]
`ip.cn`还是显示我自己本地的IP，而不是国外代理IP
`ip138.com`没问题

配置试了很多中：
如只写域名的
```
ip.cn
ip138.com
```
按规则写的
```
||ip.cn^
||ip138.com^
```
测试用的
```
||ip.cn^
||ip.cn
||ip138.com^
||ip138.com
```
`ip.cn`始终显示的是我本地的IP
但是，当我把ss设为全局模式时，显示的就是我在国外的代理服务器IP地址了

`fonts.gstatic.com`跟`ip.cn`是一样的结果。

**Comment by Nomia** [comment]
fonts.gstatic.com难道也在白名单中？

**Comment by skyjerry** [comment]
> `ip.cn`还是显示我自己本地的IP，而不是国外代理IP
> `ip138.com`没问题
> 
> 配置试了很多中：
> 如只写域名的
> 
> ```
> ip.cn
> ip138.com
> ```
> 按规则写的
> 
> ```
> ||ip.cn^
> ||ip138.com^
> ```
> 测试用的
> 
> ```
> ||ip.cn^
> ||ip.cn
> ||ip138.com^
> ||ip138.com
> ```
> `ip.cn`始终显示的是我本地的IP
> 但是，当我把ss设为全局模式时，显示的就是我在国外的代理服务器IP地址了
> 
> `fonts.gstatic.com`跟`ip.cn`是一样的结果。

我也遇到了同样的问题。

**Comment by hugbio** [comment]
我也遇到这问题。就是没效果。规则比如：||ip138.com^，但是就是不走代理。已经关闭软件重新开。都没用。

**Comment by TonySFU** [comment]
我也遇到了同样的问题。

**Comment by Nomia** [comment]
有些有效果，例如finance.yahoo.com
有些没效果，例如fonts.gstatic.com

**Comment by Nomia** [comment]
<img width="132" alt="screenshot" src="https://user-images.githubusercontent.com/1707818/55597456-48c7e100-5780-11e9-9d78-ab54db04e5b7.png">
finance.yahoo.com 有效果
fonts.gstatic.com 无效果

**Comment by hugbio** [comment]
> <img alt="screenshot" width="132" src="https://user-images.githubusercontent.com/1707818/55597456-48c7e100-5780-11e9-9d78-ab54db04e5b7.png">
> 
> finance.yahoo.com 有效果 fonts.google.com 有效果 fonts.gstatic.com 无效果

我已经找到问题了。是因为用户规则与pac规则冲突了。无效的规则是因为存在白名单的缘故。需要手动删，但是很不方便。更新pac规则后还是要再次手动删除

**Comment by p5ydn0** [comment]
上述所描述的方法我都试了，但都无效。所以有最终有效的解决方法吗？

macos v10.15.1 ss的pac文件和user-rule文件怎么找不到呢？

**Comment by Chuck1sn** [comment]
> > <img alt="screenshot" width="132" src="https://user-images.githubusercontent.com/1707818/55597456-48c7e100-5780-11e9-9d78-ab54db04e5b7.png">
> > finance.yahoo.com 有效果 fonts.google.com 有效果 fonts.gstatic.com 无效果
> 
> 我已经找到问题了。是因为用户规则与pac规则冲突了。无效的规则是因为存在白名单的缘故。需要手动删，但是很不方便。更新pac规则后还是要再次手动删除

@hugbio 手动删除的话应该如何删除呢？关键根本找不到本身的 pac 规则存在什么地方的啊。

**Comment by HaloWang** [comment]
建议添加一个规则测试功能

**Comment by HaloWang** [comment]
```
ip138.com
```
我试了各种规则都没有效果
```
ShadowsocksX-NG
Version 1.9.4 (1)
```
每次都要重启 ShadowsocksX-NG 才能测试效果实在是有点难受

**Comment by huangze1008** [comment]
> ```
> ip138.com
> ```
> 
> 我试了各种规则都没有效果
> 
> ```
> ShadowsocksX-NG
> Version 1.9.4 (1)
> ```
> 
> 每次都要重启 ShadowsocksX-NG 才能测试效果实在是有点难受

gfwlist里把ip138ban掉了, 直接改 ~/.ShadowsocksX-NG/gfwlist.js ,把 @@||ip138.com 那一行删了就完事儿，不用重启。切下开关就生效了
